### ISSUE TITLE
Deferred script rendering does not work well with output caching

### ISSUE THREAD

**Issue by ctrlaltca** [pradosoft/prado#411]
_From [google...@pcforum.hu](https://code.google.com/u/112363926660835410968/) on June 23, 2012 00:50:45_

The changes introduced in r3159 do not work properly under all circumstances, and in some cases, when output caching, might lead to the same script files being loaded multiple times (which is a problem, because all variables defined in those script files will be reset and all functions called or objects created upon loading of the script will be executed/created multiple times). What steps will reproduce the problem? 1. Run the testcase
2. When prompted about the need to have the page reloaded agree
3. Observe the bugcheck message What is the expected output? What do you see instead? The page generated in the 2nd run of the testcase (ie. when the content inside the TOutputCache has already been cached) should be the same as the one generated in the 1st run cycle (when the content has not been cached yet), so the control script "controlscript.js" should be loaded only once in the page. That, however, is not what happens, and "controlscript.js" will be loaded twice in the page if the content inside the TOutputCache control can be and gets retrieved from the application cache. What version of the product are you using? On what operating system? Prado/trunk Please provide any additional information below. This happens because the current clientscript manager implementation does not keep track of what scripts might have been registered inside the cached block, and thus might duplicate any script registrations that have happened (been emitted) inside cached content blocks.

I've attached a fix for the issue which corrects this behaviour by ensuring that all scripts loaded inside a cached content block will be marked as already loaded in the clientscript manager if said block is retrieved an emitted from the cache.

**Attachment:** [prado-deferred-script-rendering-and-output-caching-testcase.zip prado-deferred-script-rendering-and-output-caching-fix.zip](http://code.google.com/p/prado3/issues/detail?id=410)

_Original issue: http://code.google.com/p/prado3/issues/detail?id=410_
